INSTALLER: 16.04 and Kinetic integration.

RustyBamboo opened the pull request:
- Installs Catkin, Rosdep, binutils, Kinetic-desktop-full, OMPL, spacenav-node
- Downgrades Google's protobuf to compile Gazebo7's defaulted 2.6 protoc messages

Creating a pull request to encourage participation in developing and testing installer.

RustyBamboo committed:
INSTALLER: 16.04 and Kinetic integration.

- Installs Catkin, Rosdep, binutils, Kinetic-desktop-full, OMPL, spacenav-node
- Downgrades Google's protobuf to compile Gazebo7's defaulted 2.6 protoc messages

DSsoto commented on install.sh:
```diff
@@ -1,81 +1,155 @@
+
+    # The script will not allow the user to install on an unsupported OS
     instwarn "Terminating installation due to incorrect OS (detected $DTETCTED_OS)"
+    instwarn "MIL projects require Ubuntu 14.04 (trusty)"
```
This line is no longer applicable.

DSsoto commented on install.sh:
```diff
@@ -1,81 +1,155 @@
-sudo apt-get install -qq python-catkin-pkg python-rosdep gazebo7
+sudo apt-get install -qq catkin python-rosdep
+
+instlog "Installing ROS Kinetic base packages"
```
This is a desktop full install, not base packages. Let's not mislead people.

DSsoto commented on install.sh:
```diff
@@ -85,55 +159,92 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
+sudo apt-get install -qq ros-kinetic-ompl
+sudo apt-get install -qq ros-kinetic-spacenav-node
+
+# Get older version of Google protobuf for Gazebo
```
Have you confirmed with others that this is necessary, or the best way to do it?

RustyBamboo replied:
The gazebo messages are defaulted to protoc v2.6... The Ubuntu package is v3.0 (https://launchpad.net/ubuntu/+source/protobuf.) The different versions caused mine not to build. (Not sure if this is the case for others)

DSsoto commented on install.sh:
```diff
@@ -85,55 +159,92 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
-if (env | grep --quiet "SEMAPHORE=true"); then
-    mv ~/Sub8 "$CATKIN_DIR/src"
-fi
+CMAKE_PREFIX_PATH=/opt/ros/kinetic
```
Could this potentially overwrite something important? Wouldn't it be better to just append?

DSsoto commented on install.sh:
```diff
@@ -1,81 +1,155 @@
+}
+
+REQUIRED_OS="xenial"
+CATKIN_DIR=~/mil_ws
```
Should we prompt the user and have this as the default?

DSsoto reviewed (commented):
This is a good start. In my previous commit, I added a .sub_aliases file at the root of the repo, you should source this from the user's bashrc. I think we should also follow this configuration style for the other MIL repos.

We still need to find out how we're going to tie in CUDA into this. On machines not running already runing nvidia drivers this could be non-trivial.

RustyBamboo committed:
INSTALLER: Prints correct warning if not using Ubuntu 16.04

RustyBamboo committed:
INSTALLER: More checks, aliases, and improved echo texts

- Asks and confirms catkin directory
- Checks if protoc version is 2.6.1
- Fixed typo of check if Sub8 directory already exists
- Appends source sub_aliases to bashrc

RustyBamboo committed:
INSTALLER: Appends CMAKE_PREFIX_PATH

RustyBamboo committed:
INSTALLER: Correct aliases path

DSsoto commented on install.sh:
```diff
@@ -62,7 +62,7 @@ done
 # The paths to the aliases configuration files
 BASHRC_FILE=~/.bashrc
-ALIASES_FILE=$CATKIN_DIR/src/.sub_aliases
+ALIASES_FILE=$CATKIN_DIR/src/sub8/.sub_aliases
```
The `s` in Sub8 should be capital.

RustyBamboo committed:
INSTALLER: sub8 to Sub8

sentree commented on install.sh:
```diff
@@ -1,81 +1,166 @@
 
 instlog() {
-    printf "$GOODPREFIX $@ $NOCOLOR\n"
+    printf "$LOGPREFIX $@ $NOCOLOR\n"
```
I usually edit my bash scripts in nano, so naturally I prefer tabs to spaces. If there is a strong push to switch this to spaces, I recommend we use the bash standard of two spaces for indentation; however, I'd prefer to stick to tabs.

RustyBamboo replied:
Standard is probably better

RustyBamboo replied:
Although convenience is also good, which is probably better than following standards

sentree commented on install.sh:
```diff
@@ -1,81 +1,166 @@
+    fi
+    return `false`
+}
+
```
Is there any particular reason you decided to do away with ros_git_get? Jake wrote the function to simplify the lines where git repositories were cloned and I was planning on using it's '--recursive' option to make our dependencies git submodules.

RustyBamboo replied:
I saw no real need to add more complexity. "git clone bleh" done. Although, it may be worth to add it back for the recursive option.

sentree commented on install.sh:
```diff
@@ -1,81 +1,166 @@
+    case $yn in
+        [Yy]* ) break;;
+        [Nn]* ) read -p "Enter a new directory: " CATKIN_DIR; break;;
+        * ) echo "Please answer yes or no.";;
```
This could be confusing. Your code responds to values in [Y, y, N, n], but you request a 'yes' or 'no'.

RustyBamboo replied:
Woops.

sentree commented on install.sh:
```diff
@@ -1,81 +1,166 @@
+# Confirmation of where to put CATKIN directory
+CATKIN_DIR=~/mil_ws
+while true; do
+    instlog "The CATKIN Directory is $CATKIN_DIR. Is this okay(y/n)? "
```
Using a prompt based approach instead of script flags will probably work well for people that are new to MIL, but I will need to add a prompt for each flag in the common install script.

RustyBamboo replied:
I think it's a good thing to have

sentree commented on install.sh:
```diff
@@ -1,81 +1,166 @@
+
+# The paths to the aliases configuration files
+BASHRC_FILE=~/.bashrc
+ALIASES_FILE=$CATKIN_DIR/src/Sub8/.sub_aliases
```
It was a good idea to source the bash aliases file in the repository, but I will probably move some of these to software-common. I would also like to keep the file visible and in the scripts directory.

DSsoto replied:
Make sure you keep any specific to Sub8 in that file.

sentree replied:
That's the plan, a common aliases file and then one for each project installed.

sentree commented on install.sh:
```diff
@@ -1,81 +1,166 @@
+
+instlog "Installing the full ROS Kinetic"
+sudo apt-get install -qq ros-kinetic-desktop-full 
 
```
Good thinking here, we can finally do away with all of that metapackage nonsense!

sentree replied:
Installing only ros-indigo-desktop-full on Semaphore saved a lot of time during builds, so I will have to add that back in.

sentree commented on install.sh:
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
 fi
 rosdep update
 
+instlog "Installing Sub8 ROS dependencies"
```
These should be under 'Dependency Installation'

sentree commented on install.sh:
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
-    mv ~/Sub8 "$CATKIN_DIR/src"
-fi
+CMAKE_PREFIX_PATH=$CMAKE_PREFIX_PATH:/opt/ros/kinetic
+git clone https://github.com/ros-drivers/driver_common
```
What is this repository needed for? Also, I believe CMAKE_PREFIX_PATH is set when /opt/ros/kinetic/setup.bash is sourced.

RustyBamboo replied:
The driver_base in the drive_common is needed for the Camera1394 package.

Yeah, that seems to be the case...

sentree commented on install.sh:
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
-fi
+CMAKE_PREFIX_PATH=$CMAKE_PREFIX_PATH:/opt/ros/kinetic
+git clone https://github.com/ros-drivers/driver_common
+catkin_make -C $CATKIN_DIR -B
```
This build should only be done if the workspace has just been created, otherwise it will rebuild the user's existing workspace from scratch.

RustyBamboo replied:
Yeah... the problem was that the Camera1394 depends on the driver_base... But during compilation Camera1394 would fail to build because it was looking for driver_base dependences which didn't exist yet. No clue how to fix this... My solution was to build driver_common first. Maybe build driver_common into the /ros/kinetic

sentree commented on install.sh:
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
-fi
+mv $CATKIN_DIR/src/Sub8/gnc/sub8_trajectory_generator $CATKIN_DIR/src/Sub8/gnc/.sub8_trajectory_generator
+
+catkin_make -C $CATKIN_DIR -B
```
This line would result in building the repository twice on semaphore without a check.

RustyBamboo replied:
Semaphore is gud.

Not sure what the solution is for the camera1394 depending on driver_common without building twice...

sentree commented on install.sh:
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
+sudo apt-get install -qq python-twisted socat
+
+# Package management
+sudo pip install -q -U setuptools
```
You never install python-pip, which is not installed on Ubuntu server by default, so any pip installation would fail.

sentree commented on install.sh:
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
+    sudo rm -r protobuf-2.6.1.tar.gz
+else
+    instlog "Detected correct version of protoc"
+fi
```
Are we sure there isn't a PPA for this?

RustyBamboo replied:
The PPA is for the newer version (3.0.1 - i think.) The problem is that all the gazebo protoc messages don't specify that they're using 3.0.1, and so they default to 2.6.1.

sentree commented on install.sh:
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
+instlog "Cloning common GIT repositories"
 cd $CATKIN_DIR/src
-$CATKIN_DIR/src/Sub8/scripts/get_dependencies.sh
+git clone https://github.com/RustyBamboo/camera1394.git
```
Why is this necessary as opposed to installing ros-kinetic-camera1394?

RustyBamboo replied:
http://wiki.ros.org/camera1394... There doesn't seem to be support for kinetic...

sentree commented on install.sh:
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
+    echo "# Adds Sub8 aliases to shell environment" >> $BASHRC_FILE
+    echo "source $ALIASES_FILE" >> $BASHRC_FILE
 fi
 
```
I didn't see you move the Sub8 repository into the catkin workspace for Semaphore. If this step is not done, the master branch of Sub8 is what will be built, not the branch that was last updated.

RustyBamboo replied:
Ah... Smart.

sentree reviewed (changes requested):
Since we are moving toward a software-common repository, we are trying to build a unified MIL install script. I will be building and testing one on that repository over the next week and I will commit the changes from this PR that work well with you as the author. There are still a few questions I have before I add the changes, though.

DSsoto commented:
This was a great starting point but we have decided to move all of our install scripts to the installer repo.
Closing.
